contribs/vsp/src/main/java/playground/vsp/pt/fare/DistanceBasedPtFareHandler.java

```diff
@@ -0,0 +1,86 @@
+package playground.vsp.pt.fare;
+
+import com.google.inject.Inject;
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.TransportMode;
+import org.matsim.api.core.v01.events.ActivityStartEvent;
+import org.matsim.api.core.v01.events.PersonArrivalEvent;
+import org.matsim.api.core.v01.events.PersonDepartureEvent;
+import org.matsim.api.core.v01.events.PersonMoneyEvent;
+import org.matsim.api.core.v01.events.handler.ActivityStartEventHandler;
+import org.matsim.api.core.v01.events.handler.PersonArrivalEventHandler;
+import org.matsim.api.core.v01.events.handler.PersonDepartureEventHandler;
+import org.matsim.api.core.v01.network.Link;
+import org.matsim.api.core.v01.network.Network;
+import org.matsim.api.core.v01.population.Person;
+import org.matsim.core.api.experimental.events.EventsManager;
+import org.matsim.core.router.StageActivityTypeIdentifier;
+import org.matsim.core.utils.geometry.CoordUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DistanceBasedPtFareHandler implements PersonDepartureEventHandler, PersonArrivalEventHandler, ActivityStartEventHandler {
+    @Inject
+    private EventsManager events;
+
+    private final double baseFare;
+    private final double distanceFare;
+    private final Network network;
+
+    private final Map<Id<Person>, Id<Link>> personDepartureMap = new HashMap<>();
+    private final Map<Id<Person>, Id<Link>> personArrivalMap = new HashMap<>();
+
+    public DistanceBasedPtFareHandler(DistanceBasedPtFareParams params,Network network){
+        this.baseFare = params.getBaseFare();
+        this.distanceFare = params.getDistanceFare();
+        this.network = network;
+    }
+
+    @Override
+    public void handleEvent(PersonDepartureEvent event) {
+        if (event.getLegMode().equals(TransportMode.pt)){
+            if (!personDepartureMap.containsKey(event.getPersonId())){
+                personDepartureMap.put(event.getPersonId(), event.getLinkId()); // Departure place is fixed
+            }
+        }
+    }
+
+    @Override
+    public void handleEvent(PersonArrivalEvent event) {
+        if (event.getLegMode().equals(TransportMode.pt)){
+            personArrivalMap.put(event.getPersonId(), event.getLinkId()); // Arrival place can be updated (e.g. transfer to another PT line)
+        }
+    }
+
+    @Override
+    public void handleEvent(ActivityStartEvent event) {
+        if (!StageActivityTypeIdentifier.isStageActivity(event.getActType())){
+            Id<Person> personId = event.getPersonId();
+            if (personDepartureMap.containsKey(personId)){
+                if (!personArrivalMap.containsKey(personId)){
+                    throw new RuntimeException("A person has departed but not arrived. This should not happen! " +
+                            "Please make sure the event is read/handled in the normal sequence");
+                }
+                Link fromLink = network.getLinks().get(personDepartureMap.get(personId));
```

Comment:
Thanks for your work on that topic! The current version here seems as if it essentially only considers the last pt leg. What we want is the pt trip, i.e. the first boarding stop and the last arrival stop. This avoids the problem I explained with direct train being cheaper than a route using a change of trains (the transfer station makes for one more leg) although passing over the exact same tracks and having the same fare.